File: cpp/Blind75/TwoPointer/ContainerWithMostWater.cpp
class Solution {
    public:
        int maxArea(vector<int>& h) {
            int mw = 0;
    
            int size = h.size();
            for (int i = 0; i < size; ++i) {
                for (int j = i + 1; j < size; ++j) {
                    int wd = j - i;
                    int ht = min(h[i], h[j]);
                    mw = max(mw, wd * ht);
                }
            }
    
            return mw;
        }

        int maxAreaOpt(vector<int>& h) {
            int mw = 0;
            int size = h.size();
    
            int l = 0;
            int r = size - 1;
            while(l < r) {
                mw = max(mw, (r - l) * min(h[l], h[r]));
                if (h[l] < h[r]) {
                    ++l;
                } else {
                    --r;
                }
            }

            return mw;
        }
    };
    

File: Python/Grind75/ContainerWithMostWater.py
from typing import List

# https://leetcode.com/problems/container-with-most-water/


class Solution:
    def maxArea(self, height: List[int]) -> int:
        area = 0

        l = 0
        r = len(height) - 1

        while l < r:
            minH = min(height[l], height[r])
            width = r - l

            newArea = minH * width
            area = max(area, newArea)

            if (height[l] < height[r]):
                l += 1
            else:
                r -= 1
        return area


if __name__ == "__main__":
    height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    print(Solution().maxArea(height))
